mlir/lib/Conversion/RockToGPU/RockToGPU.cpp

```diff
@@ -119,6 +121,38 @@ struct MIIdRewritePattern : public OpRewritePattern<Tmi> {
     return success();
   }
 };
+
+struct WorkgroupIdRewritePattern
+    : public OpRewritePattern<rock::WorkgroupIdOp> {
+  using OpRewritePattern<rock::WorkgroupIdOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(rock::WorkgroupIdOp op,
+                                PatternRewriter &b) const override {
+    Location loc = op.getLoc();
+    bool isReverseGrid = false;
+    auto maybeIsReverseGrid = rock::getReverseGrid(op);
+    if (succeeded(maybeIsReverseGrid)) {
+      isReverseGrid = true;
+    }
+    if (isReverseGrid) {
+      FailureOr<IntegerAttr> maybeGridSize = rock::getGridSize(op);
+      if (failed(maybeGridSize)) {
+        return op->emitError("grid_size should ve been set by now.\n");
+      }
+      int64_t gridSize = maybeGridSize.value().getValue().getSExtValue();
+      AffineMap reverseMap = rock::getIdxReversalMap(b);
+      Value gridSizeVal = b.createOrFold<arith::ConstantIndexOp>(loc, gridSize);
+      Value blockIdVal = b.create<gpu::BlockIdOp>(loc, gpu::Dimension::x);
+      b.replaceOpWithNewOp<affine::AffineApplyOp>(
+          op, b.getIndexType(), reverseMap,
```

Comment:
... why the affine apply? Just `arith.sub` will be fine and won't complicate the lowering

Reply:
See the comment below.

Reply:
`expandAffineMap()` type calls might be the right move here?

Reply:
My vague sense that we should expand the map out in place instead of having `affine.apply` sitting around in code that's set to get translated into LLVM IR soon aside, I don't have any real objections here.